File: algorithms/1_easy/S/00821_shortest_distance_to_a_character/solution.cpp
class Solution {
public:
    static vector<int> shortestToChar(const string& s, char c) {
        const size_t len{s.size()};
        vector<int> res(len, numeric_limits<int>::max());
        int last_pos{-1};

        for (size_t i{0}; i < len; ++i) {
            if (s[i] == c) {
                last_pos = static_cast<int>(i);
            }

            if (last_pos != -1) {
                res[i] = static_cast<int>(i) - last_pos;
            }
        }

        last_pos = -1;
        for (size_t i{len - 1}; i < len; --i) {
            if (s[i] == c) {
                last_pos = static_cast<int>(i);
            }

            if (last_pos != -1) {
                res[i] = std::min(res[i], last_pos - static_cast<int>(i));
            }
        }

        return res;
    }
};

File: algorithms/1_easy/S/00821_shortest_distance_to_a_character/solution.rs
impl Solution {
    pub fn shortest_to_char(s: String, c: char) -> Vec<i32> {
        let len = s.len();
        let mut res = vec![i32::MAX; len];
        let mut last_pos = -1_i32;
        let byte_c = c as u8;

        for (i, byte) in s.as_bytes().iter().enumerate() {
            if *byte == byte_c {
                last_pos = i as i32;
            }

            if last_pos != -1 {
                res[i] = i as  i32 - last_pos;
            }
        }

        last_pos = -1;
        for (i, byte) in s.as_bytes().iter().rev().enumerate() {
            if *byte == byte_c {
                last_pos = i as i32;
            }

            if last_pos != -1 {
                res[len - 1 - i] = res[len - 1 - i].min(i as i32 - last_pos);
            }
        }

        res
    }
}
